homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,204 @@
+                self.async_update_battery_info,
+            )
+        )
+        async_dispatcher_send(
+            self.hass, f"{SONOS_ENTITY_CREATED}-{self.soco.uid}", SENSOR_DOMAIN
+        )
+
+    async def async_update_battery_info(self, event: SonosEvent = None) -> None:
+        """Update battery info using the provided SonosEvent."""
+        if event is None:
+            return
+
+        if (more_info := event.variables.get("more_info")) is None:
+            return
+
+        more_info_dict = dict(x.split(":") for x in more_info.split(","))
+        self._last_event = dt_util.utcnow()
+
+        is_charging = EVENT_CHARGING[more_info_dict["BattChg"]]
+        if is_charging == self.charging:
+            self._battery_info.update({"Level": int(more_info_dict["BattPct"])})
+        else:
+            if battery_info := await self.hass.async_add_executor_job(
+                fetch_battery_info_or_none, self.soco
+            ):
+                self._battery_info = battery_info
+
+        self.async_write_ha_state()
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID of the sensor."""
+        return f"{self.soco.uid}-battery"
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return f"{self.speaker.zone_name} Battery"
+
+    @property
+    def device_class(self) -> str:
+        """Return the entity's device class."""
+        return DEVICE_CLASS_BATTERY
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Get the unit of measurement."""
+        return PERCENTAGE
+
+    async def async_update(self, event=None) -> None:
+        """Poll the device for the current state."""
+        if not self.available:
+            # wait for the Sonos device to come back online
+            return
+
+        if (
+            self._last_event
+            and dt_util.utcnow() - self._last_event < BATTERY_SCAN_INTERVAL
+        ):
+            return
+
+        if battery_info := await self.hass.async_add_executor_job(
+            fetch_battery_info_or_none, self.soco
+        ):
+            self._battery_info = battery_info
+            self.async_write_ha_state()
+
+    @property
+    def battery_level(self) -> int:
+        """Return the battery level."""
+        return self._battery_info.get("Level", 0)
+
+    @property
+    def power_source(self) -> str:
+        """Return the name of the power source.
+
+        Observed to be either BATTERY or SONOS_CHARGING_RING or USB_POWER.
+        """
+        return self._battery_info.get("PowerSource", STATE_UNKNOWN)
+
+    @property
+    def charging(self) -> bool:
+        """Return the charging status of this battery."""
+        return self.power_source not in ("BATTERY", STATE_UNKNOWN)
+
+    @property
+    def icon(self) -> str:
+        """Return the icon of the sensor."""
+        return icon_for_battery_level(self.battery_level, self.charging)
```

Comment:
We normally don't need to set a custom icon for battery sensors with that device class.

Reply:
I'm not finding how this is handled automatically in the core and this is how several other integrations approach the battery icons today. Is there a newer frontend feature that takes care of this based on device class?

Reply:
Looks like this is where it's handled now: https://github.com/epenet/home-assistant-frontend/blob/dev/src/common/entity/battery_icon.ts

Reply:
The frontend will take care of it based on device class, yes.